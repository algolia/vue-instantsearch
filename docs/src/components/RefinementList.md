---
title: RefinementList
mainTitle: Components
layout: main.pug
category: Components
withHeadings: true
navWeight: 6
editable: true
githubSource: docs/src/components/RefinementList.md
---

A refinement list takes all possible values for a certain attribute and puts them in a list. A user is then able to pick one or multiple refinements for this attribute. It's also possible to search within the possible values of this attribute. Make sure that your attribute of choice is set up in `attributesForFaceting`, `searchable` if you want so.

<a class="btn btn-static-theme" href="stories/?selectedKind=RefinementList">🕹 try out live</a>

## Usage

```html
<ais-refinement-list :attribute="brands"></ais-refinement-list>
```

## Props

Name | Type | Default | Description | Required
---|---|---|---|---
attribute | string | | The attribute to refine on click | yes
searchable | boolean | `false` | You can also search within the options of this | no
operator | "or" / "and" | "or" | How to apply refinements | no
limit | number | 10 | Number of items to show | no
showMoreLimit | number | 20 | Number of items to show when the user clicked on “show more items” | no
showMore | boolean | false | Whether or not to have the option to load more values | no
sortBy | array / sort function | `['isRefined:desc', 'count:desc', 'name:asc']` | array or function to sort the results by | no
transformItems | `(items: object[]) => object[]` | | Function which receives the items, which will be called before displaying them. Should return a new array with the same shape as the original array. Useful for mapping over the items to transform, remove or reorder them | no
classNames | Object | | Override class names | no

## Slots

Name | Scope | Description
---|---|---
default | `{ items: Array, refine: Function, searchable: Boolean, searchForItems: Function, isFromSearch: Boolean, toggleShowMore: Function, isShowingMore: Boolean, createURL: Function, canRefine: Boolean, noRefinement: Boolean }` | Slot to override the DOM output
item | `{ value: String, label: String, count: Number, isRefined: Boolean, highlighted }` | Slot to override the DOM of a single item in the list
showMoreLabel | `{ isShowingMore: Boolean }` | Slot to override the text shown in the "show more" button
noResults | `{ query: String }` | Slot to override the text shown when "searchable" is true and there are no results

Note that if you override the default or item slot, and you still want highlighting when `searchable` is enabled, you should use [ais-highlight](./highlight) to have this feature: 

```html
<ais-refinement-list attribute="brand" searchable>
  <template name="item" slot-scope="{ item }">
    <ais-highlight attribute="item" :hit="item"/>
  </template>
<ais-refinement-list>
```

## CSS classes

Here's a list of CSS classes exposed by this widget. To better understand the underlying
DOM structure, have a look at the generated DOM in your browser.

Note that you can pass the prop `class-names`, with an object of class names and their replacement to override this.

Class name | Description
---|---
`ais-RefinementList` | Container class
`ais-RefinementList--noRefinement` | Container, if no refinement is possible
`ais-RefinementList-list` | The list of possible refinements
`ais-RefinementList-item` | A refinement item
`ais-RefinementList-label` | A refinement item's label
`ais-RefinementList-labelText` | A refinement item's label text
`ais-RefinementList-checkbox` | A refinement item's checkbox
`ais-RefinementList-count` | A refinement item's number of matches
`ais-RefinementList-item--selected` | Selected item
`ais-RefinementList-searchBox` | The search box when `searchable` is true
`ais-RefinementList-noResults` | In place of the refinement options when the search for facet values didn't return any values
`ais-RefinementList-showMore` | The show more button when `show-more` is true
`ais-RefinementList-showMore--disabled` | The show more button when `show-more` is true and it's not possible to be clicked

