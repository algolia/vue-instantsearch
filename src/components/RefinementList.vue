<template>
  <div :class="suit()">
    <slot v-bind="state">
      <button @click="state.refine('hi')" :class="suit('button')">example refine</button>
      <!-- ‚¨á use this to dynamically debug the state, remove it when done -->
      <json-tree :level="2" :data="state"></json-tree>
    </slot>
  </div>
</template>

<script>
import JsonTree from 'vue-json-tree'; // üëà When done, remove this
import algoliaComponent from '../component';
import { connectRefinementList } from 'instantsearch.js/es/connectors';


export default {
  // ‚¨áÔ∏è this is to help you debugging what's in `state`
  // remove it before pushing the component
  components: { 'json-tree': JsonTree },
  mixins: [algoliaComponent],
  // ‚¨áÔ∏è Those are all the options of your widget (attribute, items ...)
  // You don't need to write down the props that will be forwarded by the connector on render,
  // They are directly accessible in the state in template
  props: {
    someProp: {
      type: Array,
      required: false,
      default: () => [],
    },
  },
  data() {
    return {
      widgetName: 'RefinementList',
    };
  },
  beforeCreate() {
    this.connector = connectRefinementList;
  },
  computed: {
    // ‚¨áÔ∏è Those are all the options of your widget (attribute, items ...)
    // Same as props, just do the mapping
    widgetParams() {
      return {
        someProp: this.someProp,
      };
    },
  },
};</script>
