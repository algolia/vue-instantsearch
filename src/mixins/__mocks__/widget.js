let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
