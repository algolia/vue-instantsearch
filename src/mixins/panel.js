import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
