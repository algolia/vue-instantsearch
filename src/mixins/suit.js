import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
