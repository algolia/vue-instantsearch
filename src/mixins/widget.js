const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
