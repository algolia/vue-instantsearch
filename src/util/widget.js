export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
