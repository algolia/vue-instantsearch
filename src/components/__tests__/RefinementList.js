import { mount } from '@vue/test-utils';
import RefinementList from '../RefinementList.vue';
import { __setState } from '../../mixins/widget';

jest.unmock('instantsearch.js/es');
jest.mock('../../mixins/widget');
jest.mock('../../mixins/panel');

const defaultState = {
  items: [
    {
      value: 'yo',
      label: 'yo',
      highlighted: 'y<mark>o</mark>',
      isRefined: false,
      count: 20,
    },
    {
      value: 'how',
      label: 'how',
      highlighted: 'how',
      isRefined: false,
      count: 10,
    },
    {
      value: 'are',
      label: 'are',
      highlighted: 'are',
      isRefined: false,
      count: 8,
    },
    {
      value: 'you',
      label: 'you',
      highlighted: 'you',
      isRefined: false,
      count: 9,
    },
    {
      value: 'doing',
      label: 'doing',
      highlighted: 'doing',
      isRefined: false,
      count: 100,
    },
    { value: '?', label: '?', highlighted: '?', isRefined: false, count: 0 },
  ],
  canRefine: true,
};

it('renders correctly', () => {
  __setState({
    ...defaultState,
  });
  const wrapper = mount(RefinementList, {
    propsData: {
      attribute: 'something',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly (empty)', () => {
  __setState({
    items: [],
  });
  const wrapper = mount(RefinementList, {
    propsData: {
      attribute: 'something',
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

it("renders correctly when it's searchable", () => {
  __setState({
    ...defaultState,
  });
  const wrapper = mount(RefinementList, {
    propsData: {
      searchable: true,
      attribute: 'something',
    },
  });
  expect(wrapper.html()).toMatchSnapshot();

  expect(wrapper.find('.ais-SearchBox-input').exists()).toBe(true);
});

it("allows override of placeholder when it's searchable", () => {
  __setState({
    ...defaultState,
    searchable: true,
  });
  const placeholder = 'search in dogs';
  const wrapper = mount(RefinementList, {
    propsData: {
      searchable: true,
      attribute: 'something',
      searchablePlaceholder: placeholder,
    },
  });

  expect(wrapper.find('.ais-SearchBox-input').attributes('placeholder')).toBe(
    placeholder
  );
});

it("allows search bar classes override when it's searchable", () => {
  __setState({
    ...defaultState,
  });
  const wrapper = mount(RefinementList, {
    propsData: {
      searchable: true,
      attribute: 'something',
      classNames: {
        'ais-SearchBox-form': 'moar-classes',
      },
    },
  });
  expect(wrapper.html()).toMatchSnapshot();

  expect(wrapper.find('.ais-SearchBox-form').classes('moar-classes')).toBe(
    true
  );
});

it("disables show more if can't refine", () => {
  __setState({
    ...defaultState,
    canToggleShowMore: false,
  });
  const wrapper = mount(RefinementList, {
    propsData: {
      attribute: 'something',
      showMore: true,
    },
  });

  const showMore = wrapper.find('.ais-RefinementList-showMore');

  expect(showMore.attributes().disabled).toBe('disabled');
  expect(showMore.classes()).toContain('ais-RefinementList-showMore--disabled');

  wrapper.setData({ state: { canToggleShowMore: true } });
  expect(showMore.attributes().disabled).toBeUndefined();
  expect(showMore.classes()).not.toContain(
    'ais-RefinementList-showMore--disabled'
  );
});

it('behaves correctly', () => {
  __setState({
    ...defaultState,
    refine: jest.fn(),
  });
  const wrapper = mount(RefinementList, {
    propsData: {
      attribute: 'something',
    },
  });
  const button = wrapper.find('input[type="checkbox"]');
  button.trigger('click');
  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith('yo');
});

it('calls the Panel mixin with `canRefine`', () => {
  __setState({ ...defaultState });

  const wrapper = mount(RefinementList, {
    propsData: { attribute: 'something' },
  });

  const mapStateToCanRefine = () =>
    wrapper.vm.mapStateToCanRefine(wrapper.vm.state);

  expect(mapStateToCanRefine()).toBe(true);

  wrapper.setData({
    state: {
      canRefine: false,
    },
  });

  expect(mapStateToCanRefine()).toBe(false);

  expect(wrapper.vm.mapStateToCanRefine({})).toBe(false);
});
