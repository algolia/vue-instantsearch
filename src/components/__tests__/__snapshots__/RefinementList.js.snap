// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly (empty) 1`] = `

<div class="ais-RefinementList ais-RefinementList--noRefinement">
  <ul class="ais-RefinementList-list">
  </ul>
</div>

`;

exports[`renders correctly 1`] = `

<div class="ais-RefinementList ais-RefinementList--noRefinement">
  <ul class="ais-RefinementList-list">
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="yo"
        >
        <span class="ais-RefinementList-labelText">
          yo
        </span>
        <span class="ais-RefinementList-count">
          20
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="how"
        >
        <span class="ais-RefinementList-labelText">
          how
        </span>
        <span class="ais-RefinementList-count">
          10
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="are"
        >
        <span class="ais-RefinementList-labelText">
          are
        </span>
        <span class="ais-RefinementList-count">
          8
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="you"
        >
        <span class="ais-RefinementList-labelText">
          you
        </span>
        <span class="ais-RefinementList-count">
          9
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="doing"
        >
        <span class="ais-RefinementList-labelText">
          doing
        </span>
        <span class="ais-RefinementList-count">
          100
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="?"
        >
        <span class="ais-RefinementList-labelText">
          ?
        </span>
        <span class="ais-RefinementList-count">
          0
        </span>
      </label>
    </li>
  </ul>
</div>

`;

exports[`renders correctly when it's searchable 1`] = `

<div class="ais-RefinementList ais-RefinementList--noRefinement">
  <ul class="ais-RefinementList-list">
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="yo"
        >
        <span class="ais-RefinementList-labelText">
          yo
        </span>
        <span class="ais-RefinementList-count">
          20
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="how"
        >
        <span class="ais-RefinementList-labelText">
          how
        </span>
        <span class="ais-RefinementList-count">
          10
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="are"
        >
        <span class="ais-RefinementList-labelText">
          are
        </span>
        <span class="ais-RefinementList-count">
          8
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="you"
        >
        <span class="ais-RefinementList-labelText">
          you
        </span>
        <span class="ais-RefinementList-count">
          9
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="doing"
        >
        <span class="ais-RefinementList-labelText">
          doing
        </span>
        <span class="ais-RefinementList-count">
          100
        </span>
      </label>
    </li>
    <li class="ais-RefinementList-item">
      <label class="ais-RefinementList-label">
        <input type="checkbox"
               class="ais-RefinementList-checkbox"
               value="?"
        >
        <span class="ais-RefinementList-labelText">
          ?
        </span>
        <span class="ais-RefinementList-count">
          0
        </span>
      </label>
    </li>
  </ul>
</div>

`;
